src/data/finite/basic.lean

```diff
@@ -0,0 +1,316 @@
+instance because we want `fintype` instances to be useful for computations. -/
+def fintype.of_finite (α : Type*) [finite α] : fintype α :=
+fintype.of_equiv _ (finite.equiv_fin α).symm
+
+lemma finite_iff_nonempty_fintype (α : Type*) :
+  finite α ↔ nonempty (fintype α) :=
+⟨λ _, by exactI ⟨fintype.of_finite α⟩, λ ⟨_⟩, by exactI infer_instance⟩
+
+lemma finite_or_infinite (α : Type*) :
+  finite α ∨ infinite α :=
+begin
+  casesI fintype_or_infinite α,
+  { exact or.inl infer_instance },
+  { exact or.inr infer_instance }
+end
+
+lemma not_finite (α : Type*) [h1 : infinite α] [h2 : finite α] : false :=
+by { haveI := fintype.of_finite α, exact not_fintype α }
+
+lemma finite.of_not_infinite {α : Type*} (h : ¬ infinite α) : finite α :=
+finite.of_fintype (fintype_of_not_infinite h)
+
+lemma infinite.of_not_finite {α : Type*} (h : ¬ finite α) : infinite α :=
+⟨λ h', h (finite.of_fintype h')⟩
+
+lemma not_infinite_iff_finite {α : Type*} : ¬ infinite α ↔ finite α :=
+⟨finite.of_not_infinite, λ h h', by exactI not_finite α⟩
+
+lemma not_finite_iff_infinite {α : Type*} : ¬ finite α ↔ infinite α :=
+not_infinite_iff_finite.not_right.symm
+
+lemma nat.card_eq (α : Type*) :
+  nat.card α = if h : finite α then by exactI @fintype.card α (fintype.of_finite α) else 0 :=
+begin
+  casesI finite_or_infinite α,
+  { haveI := fintype.of_finite α,
+    simp only [*, nat.card_eq_fintype_card, dif_pos],
+    congr, },
+  { simp [*, not_finite_iff_infinite.mpr h] },
+end
+
+lemma finite.card_pos_iff [finite α] :
+  0 < nat.card α ↔ nonempty α :=
+begin
+  haveI := fintype.of_finite α,
+  simp only [nat.card_eq_fintype_card],
+  exact fintype.card_pos_iff,
+end
+
+@[nolint instance_priority]
+instance finite.prop (p : Prop) : finite p :=
+if h : p
+then ⟨(equiv.prop_equiv_punit h).trans (by simpa using fintype.equiv_fin punit)⟩
+else ⟨(equiv.prop_equiv_pempty h).trans (by simpa using fintype.equiv_fin pempty)⟩
+
+namespace finite
+
+lemma exists_max [finite α] [nonempty α] [linear_order β] (f : α → β) :
+  ∃ x₀ : α, ∀ x, f x ≤ f x₀ :=
+by { haveI := fintype.of_finite α, exact fintype.exists_max f }
+
+lemma exists_min [finite α] [nonempty α] [linear_order β] (f : α → β) :
+  ∃ x₀ : α, ∀ x, f x₀ ≤ f x :=
+by { haveI := fintype.of_finite α, exact fintype.exists_min f }
+
+instance {α : Sort*} [finite α] : finite (plift α) := finite.of_equiv _ equiv.plift.symm
+
+lemma of_bijective {α β : Sort*} [finite α] (f : α → β) (H : function.bijective f) : finite β :=
+finite.of_equiv _ (equiv.of_bijective _ H)
+
+lemma of_surjective {α β : Sort*} [finite α] (f : α → β) (H : function.surjective f) : finite β :=
+begin
+  haveI := fintype.of_finite (plift α),
+  rw [← equiv.surjective_comp equiv.plift f, ← equiv.comp_surjective _ equiv.plift.symm] at H,
+  haveI := fintype.of_surjective _ H,
+  exact finite.of_equiv _ equiv.plift,
+end
+
+lemma of_injective {α β : Sort*} [finite β] (f : α → β) (H : function.injective f) : finite α :=
```

Comment:
If you prove this before `of_surjective`, then you can use `function.injective_surj_inv`.